### ISSUE TITLE
Refactor proposal and possible future enhancements

### ISSUE THREAD

**Issue by vigreco** [soldair/node-qrcode#56]
I open this issue to discuss about a possible re-factor of the lib.

The main reason that led me to propose this is because I found quite hard right now add tests or new functionalities to the lib in a clean way.

I started adding unit tests in my local branch but it's hard to follow what tests are actually testing.

I've also some ideas (mostly regarding the rendering part) that I would like to share and see implemented here but since most of the logic and rendering part is all contained in a couple of files, add new code there will lead at some point to more confusion and to a code harder to manage.

So, I propose these change:
#### Split qrcode.js file in smaller modules

Having small, simple modules each with its own task, will make much easier add tests, documentation and new functionalities to them, and will make code easier to follow.
Add support for other qrcode types would be also easier.
#### Remove duplicated code and organize renderers

Currently, most of the code that is used to draw the qrcode is duplicated across various functions and it's hard to test.
A more 'composable' structure could help to organize the various renderers.
We could have a **base-renderer** class that will contains all the common methods and then a list of possible renderer.

The file structure so far could be something like:

```
lib/
  core/
    8bit-byte.js
    alignment-pattern.js
    bit-buffer.js
    ecc.js
    finder-pattern.js
    mask-pattern.js
    math.js
    modes.js
    polynomial.js
    rs-blocks.js
    utils.js
    version.js
  renderer/
    base-renderer.js
    png-renderer.js
    jpg-renderer.js
    svg-renderer.js
    pdf-renderer.js
    canvas-renderer.js
    terminal-renderer.js
  qrcode.js
```
#### Add linter and code coverage

Since one of the main reason of this refactor is to promote the add of tests, having code coverage would also be helpful.
Also, enforcing code style with a linter will prevent common bugs and style inconsistency across files.
I would suggest eslint with something like [Standard JS](https://github.com/feross/standard).
#### (Future work) Remove node-canvas in place of image libs

I think _node-canvas_ is a bit overkill if used to just export images.
Renderer like png-renderer.js and jpg-renderer.js could take advantages of dedicated image libs like [pngjs](https://github.com/lukeapage/pngjs) and [jpeg-js](https://github.com/eugeneware/jpeg-js)
#### (Future work) Completely decouple logic from render

Currently the qrcode.js lib, takes care of generating the qrcode data and rendering it as an array of bit representing each modules.
This array is then used to draw the qrcode and export it as image.
This works perfectly fine, but I would like to abstract a little more this process and split the drawing methods in smaller parts.

The idea is to move all the function that do something related to how to render in (for example) base-renderer.js and let core library to just produce information (such as coords) about what and where to render.
**base-renderer** would have dedicated functions to render for example: qrcode background, finding pattern, alignment pattern, each single modules ecc.
These functions could be then extended/overriden to add support for _effects_, such as colors or deformations to allow qrcode customization (like [this](https://www.unitag.io/qrcode))
#### (Future work) Add support for promise

I think Promises support would also be a good addition to the lib.

What do you think?

**Comment by soldair** [comment]
i agree with all of this for the version 1.0 except moving to promises.
lets push that off for later. I am personally still agianst promises for a lot of reasons that are not important to this thread.

**Comment by vigreco** [comment]
Great! Thanks for your feedback. I'll start working on this asap.

Regarding Promises, I meant to add support for them in addition to the current interface.
Something like `if callback function is undefined, return a Promise`.
However this was just some preliminary thought, nothing actually needed.

**Comment by vigreco** [comment]
I've started pushing my progress on [refactor/core](https://github.com/soldair/node-qrcode/tree/refactor/core) branch.
I'm focusing on the core part for now. I hope to move to the rendering part soon. 😃

**Comment by keeslinp** [comment]
Hey, if you guys are planning on replacing node-canvas with jimp because all it does is export an image would it be possible to add a way to install this without node-canvas? The reason I ask is because all I need is the bit array and I'm manually creating a PNG myself later so the errors about not having cairo are inconvenient.

Thanks for you guys's great work :)

**Comment by vigreco** [comment]
Hi @keeslinp, I'm proposing a refactor which also removes node-canvas dependency.
You can follow the progress here: #67
Feel free to clone `refactor/renderer` branch if you want to try it.

**Comment by keeslinp** [comment]
Awesome, thanks :). I was able to get https://github.com/alexeyten/qr-image to work for now. But I think there might be some browser compatibility issues (just judging by the open issues). I'll keep an eye on this and give it a try. Thanks for the fast reply!

**Comment by soldair** [comment]
node canvas has been removed as of 0.8.0.

whats next =)

**Comment by vigreco** [comment]
All the proposed features have been implemented and are available in latest releases! 🎉 
Thanks!
